### DISCUSSION TITLE
Update with same value inside a splitted atom triggers an array update

### DISCUSSION THREAD

**Discussion by arnodb** [pmndrs/jotai#2079] [Bug report]
Hello folks,

I encountered a tiny issue where:

* I have an atom of array (of strings)
* I split it with `splitAtom`

If eventually there is an update of one of the elements where this element actually remains the same (`===` comparison returns true) then the code of `splitAtom` triggers an update of the array with a different array but with the same content.

The code I am pointing to is https://github.com/pmndrs/jotai/blob/main/src/vanilla/utils/splitAtom.ts#L114 where

* `nextItem` is computed
* regardless of its value, `set` is called with a new array

Points to consider:

* One option is to compare `nextItem` with `arr[index]` and NOT call `set` on identical values.
* One option is to compare `nextItem` with `arr[index]` and call `set` with the same array when items are identical, this cannot harm (not more then calling `set` with a different array and same content).
* One more invasive option would be to say that `set` should detect that, but I am not sure it is sensible to go that way.
* A fourth option would be to say that this is not a problem and the upstream item update should be fixed according to whatever the needs are.

I am actually using the 4th option workaround at this time as it "works". But I wanted to share the information with you.

Last, if you eventually want a PR with one of the first 2 options implemented, I can easily do it.

**Comment by dai-shi** [comment]
Nice catch. We should bail out if it's the same value. (= not call set)
Let's use `Object.is` instead of `===`.
Would you like to draft a PR? We should have a failing test in advance.

**Reply by arnodb** [reply]
> Would you like to draft a PR? We should have a failing test in advance.

If you offer help for the test then yes, please. :)

**Reply by dai-shi** [reply]
Please see how current tests are implemented.

**Reply by arnodb** [reply]
Hello, here is a first attempt: https://github.com/pmndrs/jotai/pull/2088.
